server/src/main/java/org/apache/calcite/sql/ddl/SqlCreateTable.java

```diff
@@ -176,6 +176,9 @@ public void execute(CalcitePrepare.Context context) {
             type = typeEntry.getType().apply(typeFactory);
           }
         }
+        if (type != null && !type.isNullable() && d.strategy == ColumnStrategy.NULLABLE) {
```

Comment:
I agree this is not the cleanest way to get nullable struct types, but as you say, at the point when the struct type is created, the nullability info does not get propagated. So, I think this is the simplest solution, because it does not require any side changes.
I have committed a new version that relies on `d.dataType.nullable` instead of `d.strategy` to set the correct nullability

Reply:
I don't see a better option at the moment so I agree with you.